/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

syntax = "proto3";
package proto;

option java_package = "org.apache.pulsar.functions.proto";
option java_outer_classname = "Function";

message FunctionConfig {

    enum ProcessingGuarantees {
        ATMOST_ONCE = 0;
        ATLEAST_ONCE = 1;
    }
    string tenant = 1;
    string namespace = 2;
    string name = 3;
    string className = 4;
    // map from input topic name to serde
    map<string, string> inputs = 5;
    string outputSerdeClassName = 6;
    string sinkTopic = 7;
    ProcessingGuarantees processingGuarantees = 9;
    map<string,string> userConfig = 10;
}

message PackageLocationMetaData {
    string packagePath = 1;
}

message FunctionMetaData {
    FunctionConfig functionConfig = 1;
    PackageLocationMetaData packageLocation = 2;
    string runtime = 3;
    uint64 version = 4;
    uint64 createTime = 5;
    string workerId = 6;
}

message Snapshot {
    repeated FunctionMetaData functionMetaDataList = 1;
    bytes lastAppliedMessageId = 2;

}